File: nodecore-cli/src/main/java/nodecore/cli/serialization/PoPTransactionProblemReport.java
// VeriBlock NodeCore
// Copyright 2017-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package nodecore.cli.serialization;

import nodecore.api.grpc.VeriBlockMessages;
import nodecore.api.grpc.utilities.ByteStringAddressUtility;
import org.veriblock.core.utilities.Utility;

public class PoPTransactionProblemReport {
    PoPTransactionProblemReport(VeriBlockMessages.PoPTransactionProblemReport report) {
        if (report.getAddress() != null) {
            address = ByteStringAddressUtility.parseProperAddressTypeAutomatically(report.getAddress());
        }

        if (report.getTxid() != null) {
            txid = Utility.bytesToHex(report.getTxid().toByteArray());
        }

        if (report.getEndorsedVbkBlockHash() != null) {
            endorsedVBKBlockHash = Utility.bytesToHex(report.getEndorsedVbkBlockHash().toByteArray());
        }

        endorsedVBKBlockIndex = report.getEndorsedVbkBlockIndex();

        if (report.getIncludedInVbkBlockHash() != null) {
            includedInVBKBlockHash = Utility.bytesToHex(report.getIncludedInVbkBlockHash().toByteArray());
        }

        includedInVBKBlockIndex = report.getIncludedInVbkBlockIndex();

        if (report.getPopPayoutVbkBlockHash() != null) {
            PotentialPoPPayoutVBKBlockHash = Utility.bytesToHex(report.getPopPayoutVbkBlockHash().toByteArray());
        }

        PotentialPoPPayoutVBKBlockIndex = report.getPopPayoutVbkBlockIndex();

        paidOutInPoPVBKBlock = report.getPaidOutInPopPayoutVbkBlock();

        PoPPayoutVBKAmount = Utility.formatAtomicLongWithDecimal(report.getPopPayoutVbkAmount());

        bitcoinTxID = Utility.bytesToHex(report.getBitcoinTxid().toByteArray());

        if (report.getIncludedInBtcBlockHash() != null) {
            includedInBTCBlockHash = Utility.bytesToHex(report.getIncludedInBtcBlockHash().toByteArray());
        }

        includedInBTCBlockIndex = report.getIncludedInBtcBlockIndex();

        endorsedVBKBlockInMainChain = report.getEndorsedVbkBlockInMainChain();

        includedInVBKBlockInMainChain = report.getIncludedInVbkBlockInMainChain();

        includedInBTCBlockInMainChain = report.getIncludedInBtcBlockInMainChain();

        firstBTCBlockOfEquivalentPOPEndorsementsIndex = report.getFirstBtcBlockOfEquivalentPopEndorsementsIndex();

        if (!endorsedVBKBlockInMainChain) {
            potentialCause += "This PoP transaction does not endorse a VBK block in the main VBK blockchain!\n";
        }

        if (!includedInVBKBlockInMainChain) {
            potentialCause += "This PoP transaction was not included in the main VBK blockchain!\n";
        }

        if (!includedInBTCBlockInMainChain) {
            potentialCause += "The BTC transaction was not included in a Bitcoin block in the main Bitcoin blockchain!\n";
        }

        int differenceFromBestCompetitor = firstBTCBlockOfEquivalentPOPEndorsementsIndex - includedInBTCBlockIndex;
        if (differenceFromBestCompetitor > 4) {
            potentialCause += "The BTC transaction was included " + differenceFromBestCompetitor + " BTC blocks after the first BTC block to contain a PoP publication of block " + endorsedVBKBlockHash + ".\n";
        }

        if (PotentialPoPPayoutVBKBlockIndex > report.getCurrentBlockHeight()) {
            potentialCause += "If rewarded, this PoP transaction would receive payment in VBK block " + PotentialPoPPayoutVBKBlockIndex + ", which has not occurred yet!\n";
        }

        if (includedInVBKBlockHash == null) {
            potentialCause += "The PoP transaction does not appear to have been included in any VeriBlock block!\n";
        }

        currentBlockHeight = report.getCurrentBlockHeight();

        potentialCause = potentialCause.trim();
    }

    public String address;

    public String txid;

    public String endorsedVBKBlockHash;

    public int endorsedVBKBlockIndex;

    public String includedInVBKBlockHash;

    public int includedInVBKBlockIndex;

    public String PotentialPoPPayoutVBKBlockHash;

    public int PotentialPoPPayoutVBKBlockIndex;

    public boolean paidOutInPoPVBKBlock;

    public String PoPPayoutVBKAmount;

    public String bitcoinTxID;

    public String includedInBTCBlockHash;

    public int includedInBTCBlockIndex;

    public boolean endorsedVBKBlockInMainChain;

    public boolean includedInVBKBlockInMainChain;

    public boolean includedInBTCBlockInMainChain;

    public int firstBTCBlockOfEquivalentPOPEndorsementsIndex;

    public int currentBlockHeight;

    public String potentialCause = "";
}

File: nodecore-cli/src/main/java/nodecore/cli/serialization/PopTransactionProblemReport.kt
// VeriBlock NodeCore
// Copyright 2017-2021 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package nodecore.cli.serialization

import nodecore.api.grpc.RpcPopTransactionProblemReport
import org.veriblock.sdk.extensions.toHex
import org.veriblock.sdk.extensions.toProperAddressType
import org.veriblock.core.utilities.extensions.formatAtomicLongWithDecimal
import org.veriblock.core.utilities.extensions.toHex

class PopTransactionProblemReport(
    report: RpcPopTransactionProblemReport
) {
    val address = report.address?.toProperAddressType()

    val txid = report.txid?.toHex()

    val endorsedVbkBlockHash = report.endorsedVbkBlockHash?.toHex()

    val endorsedVbkBlockIndex = report.endorsedVbkBlockIndex

    val includedInVbkBlockHash = report.includedInVbkBlockHash?.toHex()

    val includedInVbkBlockIndex = report.includedInVbkBlockIndex

    val potentialPopPayoutVbkBlockHash = report.popPayoutVbkBlockHash?.toHex()

    val potentialPopPayoutVbkBlockIndex = report.popPayoutVbkBlockIndex

    val paidOutInPopVbkBlock = report.paidOutInPopPayoutVbkBlock

    val popPayoutVbkAmount = report.popPayoutVbkAmount.formatAtomicLongWithDecimal()

    val bitcoinTxId = report.bitcoinTxid.toByteArray().toHex()

    val includedInBtcBlockHash = report.includedInBtcBlockHash?.toHex()

    val includedInBtcBlockIndex = report.includedInBtcBlockIndex

    val endorsedVbkBlockInMainChain = report.endorsedVbkBlockInMainChain

    val includedInVbkBlockInMainChain = report.includedInVbkBlockInMainChain

    val includedInBtcBlockInMainChain = report.includedInBtcBlockInMainChain

    val firstBtcBlockOfEquivalentPopEndorsementsIndex = report.firstBtcBlockOfEquivalentPopEndorsementsIndex

    val currentBlockHeight = report.currentBlockHeight

    var potentialCause = ""

    init {
        if (!endorsedVbkBlockInMainChain) {
            potentialCause += "This PoP transaction does not endorse a VBK block in the main VBK blockchain!\n"
        }
        if (!includedInVbkBlockInMainChain) {
            potentialCause += "This PoP transaction was not included in the main VBK blockchain!\n"
        }
        if (!includedInBtcBlockInMainChain) {
            potentialCause += "The BTC transaction was not included in a Bitcoin block in the main Bitcoin blockchain!\n"
        }
        val differenceFromBestCompetitor = firstBtcBlockOfEquivalentPopEndorsementsIndex - includedInBtcBlockIndex
        if (differenceFromBestCompetitor > 4) {
            potentialCause += "The BTC transaction was included $differenceFromBestCompetitor BTC blocks after the first BTC block to contain a PoP publication of block $endorsedVbkBlockHash.\n"
        }
        if (potentialPopPayoutVbkBlockIndex > report.currentBlockHeight) {
            potentialCause += "If rewarded, this PoP transaction would receive payment in VBK block $potentialPopPayoutVbkBlockIndex, which has not occurred yet!\n"
        }
        if (includedInVbkBlockHash == null) {
            potentialCause += "The PoP transaction does not appear to have been included in any VeriBlock block!\n"
        }
        potentialCause = potentialCause.trim()
    }
}
